transport.go

```diff
@@ -151,15 +155,28 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
 	if err != nil {
 		return nil, err
 	}
+
 	remoteMultiaddr, err := toQuicMultiaddr(addr)
 	if err != nil {
 		return nil, err
 	}
 	tlsConf, keyCh := t.identity.ConfigForPeer(p)
+
+	if simConnect, _ := in.GetSimultaneousConnect(ctx); simConnect {
+		if bytes.Compare([]byte(t.localPeer), []byte(p)) < 0 {
+			err = t.holePunch(network, addr)
+			if err == nil {
+				err = errors.New("hole punching attempted; no active dial")
```

Comment:
Why is this an error?

Reply:
So this is the tricky part.

Ideally we would return the connection accepted by listen, but this is problematic:
1. we need to hook into the listen and intercept that
2. the connection would be returned to the swarm twice, which is very undesirable for obvious reasons.
3. We also need to signal to the swarm that we are done trying to connect, and propagate this to the caller.

I don't think it is safe to return `nil, nil` here to the swarm, as it will probably crash if it doesn't get a connection and no error.

So I couldn't find anything better to do, so just return an error and let the caller (who actually supplied the context option after all) check for the incoming connection.

Any ideas for improving this situation are welcome!

Reply:
Granted, this should probably be a named error (if we decide to keep this logic) so that the caller knows to wait for the incoming connection and not have to guess.

Reply:
Actually, now that I think about it, if the incoming connection succeeds the `Connect` call will not return an error to the caller.
I still think we need to return an error to the swarm (so that it doesn't expect a connect) though.

Reply:
Can you point me to where this `ctx` is configured and `Dial` is called? I tried grepping through the libp2p workspace, but the only occurrence for `WithSimultaneousConnect` is in go-libp2p-core, so I assume it's only in a PR at the moment?

Reply:
It's brand new -- it's gonna come from this pr: https://github.com/libp2p/go-libp2p/pull/1057
I don't think aarsh has added yet, but he will do very soon.

Reply:
I'm ok with returning a named error here. Hooking into `Listen` and catching the connection there might be another (more complicated) solution, but I'm not sure if we're going to run into problems if both peers think that they're a client.

Reply:
Yeah, I am not sure either but I think that the swarm is not going to like it.
I'll make a named error for this, maybe it should live in core however?

Reply:
added named error.

Reply:
@vyzo 

```
Actually, now that I think about it, if the incoming connection succeeds the Connect call will not return an error to the caller.
```

Is there some magic in Swarm that does NOT return an error for the `Connect` call even if the dial fails but an incoming succeeds ? Please can you point me to it ?